### PR TITLE
GH-2534: enforce formated code and type checking in ci-cd

### PR DESCRIPTION
**Pull request by helpmefindaname**
Closes [GH-2534](https://github.com/flairNLP/flair/issues/2534) improving the CI-CD system

- [x] fix warnings
- [x] unify requirements to `requirements-dev.txt`
- [x] ~~try out https://github.com/pytest-dev/pytest-xdist for parallel testing~~ turns out to have a too high memory footprint, instead I looked at the speed of tests, reduces some epochs, down sampled datasets and reduced the testing time from ~20min to ~8min
- [x] type checking in CI-CD
- [x] code formatting and code format checking

### CODE REVIEW

**Review by alanakbik — COMMENTED** [review]
Thanks so much for all these improvements! I'll do a bit of testing and likely merge soon!

File: flair/datasets/entity_linking.py [hunk 1]
```diff
@@ -52,20 +53,17 @@ def make_entity_dict(self, label_type='nel', threshold: int = 1) -> Dictionary:
...
 
                 spans = sentence.get_spans(label_type)
                 for span in spans:
                     annotation = span.tag
                     self.total_number_of_entity_mentions += 1
-                    if annotation in self.entity_occurences:
-                        self.entity_occurences[annotation] += 1
-                    else:
-                        self.entity_occurences[annotation] = 1
+                    self.entity_occurences[annotation] = 1
```
[the remark this answers is not in the document]
Comment (alanakbik) [reply flair/datasets/entity_linking.py:66]:
```suggestion
                    self.entity_occurences[annotation] += 1
```

File: flair/datasets/entity_linking.py [hunk 2]
```diff
@@ -52,20 +53,17 @@ def make_entity_dict(self, label_type='nel', threshold: int = 1) -> Dictionary:
         ent_dictionary contains all wikinames that occure at least threshold times and gives each name an ID
         """
         self.threshold = threshold
-        self.entity_occurences = {}
+        self.entity_occurences: Dict[str, int] = defaultdict(int)
```
Comment (alanakbik) [inline flair/datasets/entity_linking.py:56]:
```suggestion
        self.entity_occurences: Counter = Counter()
```

File: flair/datasets/entity_linking.py [hunk 3]
```diff
@@ -1,30 +1,31 @@
 import csv
 import logging
 import os
+from collections import defaultdict
```
Comment (alanakbik) [inline flair/datasets/entity_linking.py:4]:
```suggestion
from collections import Counter
```

File: flair/models/sequence_tagger_model.py [hunk 4]
```diff
@@ -311,27 +299,21 @@ def predict(
...
-        if label_name == None:
+        if label_name is None:
             label_name = self.tag_type
 
         with torch.no_grad():
             if not sentences:
                 return sentences
 
-            if isinstance(sentences, Sentence):
+            if not isinstance(sentences, list):
```
[the remark this answers is not in the document]
Comment (alanakbik) [reply flair/models/sequence_tagger_model.py:309]: works!

File: flair/models/regexp_tagger.py [hunk 5]
```diff
@@ -93,8 +95,10 @@ def _listify(element: object) -> list:
 
     def predict(self, sentences: Union[List[Sentence], Sentence]) -> List[Sentence]:
         """
-            Predict the given sentences according to the registered mappings.
+        Predict the given sentences according to the registered mappings.
         """
+        if not isinstance(sentences, list):
```
[the remark this answers is not in the document]
Comment (helpmefindaname) [reply flair/models/regexp_tagger.py:100]: `>>> from flair.data import Sentence
>>> s = Sentence()
>>> isinstance(s, list)
False
`
Sentence might be used as a list of tokens, but it doesn't derive from list, hence it isn't one.
To make Sentence a list (which I wouldn't) we would have to change the class definition to `class Sentence(DataPoint, list):`

[the remark this answers is not in the document]
Comment (alanakbik) [reply flair/models/regexp_tagger.py:100]: Ah great, thanks for clarifying!

File: flair/datasets/entity_linking.py [hunk 6]
```diff
@@ -53,18 +54,14 @@ def make_entity_dict(self, label_type="nel", threshold: int = 1) -> Dictionary:
...
             if not sentence.is_document_boundary:  # exclude "-DOCSTART-"-sentences
 
                 spans = sentence.get_spans(label_type)
-                for span in spans:
-                    annotation = span.tag
-                    self.total_number_of_entity_mentions += 1
-                    self.entity_occurences[annotation] = 1
-
+                self.entity_occurences.update(span.tag for span in spans)
+        self.total_number_of_entity_mentions = sum(self.entity_occurences.values())
```
Comment (alanakbik) [inline flair/datasets/entity_linking.py:64]: Even better, thanks for this!

File: pyproject.toml [hunk 7]
```diff
@@ -13,7 +13,7 @@ exclude = '''
 )
 '''
 [tool.pytest.ini_options]
-flake8-max-line-length = 120
+flake8-max-line-length = 210
```
Comment (alanakbik) [inline pyproject.toml:16]: Seems a bit much?

**Review by alanakbik — APPROVED** [review]

**Pull request merged**